File: 2015/solution/day8.go
package solution

func (Solution) Day8Part1(fn string) int {
	lines := toLines(fn)
	c1 := 0
	c2 := 0

	for _, line := range lines {
		c1 += len(line)
		line = line[1 : len(line)-1]
		decoded := 0
		for i := 0; i < len(line); {
			if line[i] == '\\' {
				if line[i+1] == '\\' || line[i+1] == '"' {
					i += 2
				} else if line[i+1] == 'x' {
					i += 4
				}
			} else {
				i++
			}
			decoded++
		}
		c2 += decoded
	}

	return c1 - c2
}

func (Solution) Day8Part2(fn string) int {
	lines := toLines(fn)
	c1 := 0
	c2 := 0

	for _, line := range lines {
		c1 += len(line)
		encoded := 0
		for i := 0; i < len(line); {
			if line[i] == '\\' || line[i] == '"' {
				encoded += 2
			} else {
				encoded++
			}
			i++
		}
		c2 += (encoded + 2)
	}

	return c2 - c1
}

File: 2016/solution/day8.py
from util import toLines


def one(filename: str):
    lines = toLines(filename)
    matrix = [[0 for _ in range(50)] for _ in range(6)]
    for line in lines:
        if line.startswith("rect"):
            x, y = map(int, line.split(" ")[1].split("x"))
            for i in range(y):
                for j in range(x):
                    matrix[i][j] = 1
        elif line.startswith("rotate row"):
            row = int(line.split(" ")[2].split("=")[1])
            amount = int(line.split(" ")[4])
            matrix[row] = matrix[row][-amount:] + matrix[row][:-amount]
        elif line.startswith("rotate column"):
            col = int(line.split(" ")[2].split("=")[1])
            amount = int(line.split(" ")[4])
            column = [matrix[i][col] for i in range(len(matrix))]
            column = column[-amount:] + column[:-amount]
            for i in range(len(matrix)):
                matrix[i][col] = column[i]

    return sum([sum(row) for row in matrix])


def two(filename: str):
    lines = toLines(filename)
    matrix = [[0 for _ in range(50)] for _ in range(6)]
    for line in lines:
        if line.startswith("rect"):
            x, y = map(int, line.split(" ")[1].split("x"))
            for i in range(y):
                for j in range(x):
                    matrix[i][j] = 1
        elif line.startswith("rotate row"):
            row = int(line.split(" ")[2].split("=")[1])
            amount = int(line.split(" ")[4])
            matrix[row] = matrix[row][-amount:] + matrix[row][:-amount]
        elif line.startswith("rotate column"):
            col = int(line.split(" ")[2].split("=")[1])
            amount = int(line.split(" ")[4])
            column = [matrix[i][col] for i in range(len(matrix))]
            column = column[-amount:] + column[:-amount]
            for i in range(len(matrix)):
                matrix[i][col] = column[i]
    for row in matrix:
        print("".join(["#" if x == 1 else " " for x in row]))


print("part one:", one("input/day8"))
print("part two:")
two("input/day8")
